### ISSUE TITLE
📉 Implement Loss Functions for Neural Network Training

### ISSUE THREAD

**Issue by sakshambedi** [sakshambedi/micrograd#19]
# 📉 Implement Loss Functions for Neural Network Training

## 📋 Description

Implement essential loss functions required for training neural networks and machine learning models. Loss functions are critical for computing gradients and optimizing model parameters during training.

## 🎯 Acceptance Criteria

- [ ] Implement `MSELoss` (Mean Squared Error) for regression tasks
- [ ] Implement `CrossEntropyLoss` for multi-class classification
- [ ] Implement `BCELoss` (Binary Cross Entropy) for binary classification
- [ ] Implement `BCEWithLogitsLoss` for numerical stability
- [ ] Add `L1Loss` (Mean Absolute Error) for robust regression
- [ ] Implement `NLLLoss` (Negative Log Likelihood) as building block
- [ ] Add reduction modes: 'mean', 'sum', 'none'
- [ ] Support for class weights and ignore_index
- [ ] Implement proper gradient computation for all losses
- [ ] Add comprehensive tests with edge cases

### Current State
- Loss functions: ❌ Not implemented
- Gradient computation: ✅ Infrastructure exists
- Reduction operations: ✅ Basic sum/mean available
- Activation functions: ❌ Need sigmoid, softmax (Issue #06)

### Implementation Requirements

**Base Loss Class**
```python
class Loss(Module):
    """Base class for all loss functions."""
    
    def __init__(self, reduction='mean'):
        super().__init__()
        self.reduction = reduction
    
    def _reduce(self, loss):
        """Apply reduction to loss tensor."""
        if self.reduction == 'mean':
            return loss.mean()
        elif self.reduction == 'sum':
            return loss.sum()
        elif self.reduction == 'none':
            return loss
        else:
            raise ValueError(f"Invalid reduction: {self.reduction}")
```

**Mean Squared Error Loss**
```python
class MSELoss(Loss):
    """Mean Squared Error loss for regression."""
    
    def forward(self, input, target):
        diff = input - target
        loss = diff ** 2
        return self._reduce(loss)
```

**Cross Entropy Loss**
```python
class CrossEntropyLoss(Loss):
    """Cross-entropy loss for multi-class classification."""
    
    def __init__(self, weight=None, ignore_index=-100, reduction='mean'):
        super().__init__(reduction)
        self.weight = weight
        self.ignore_index = ignore_index
    
    def forward(self, input, target):
        # input: (N, C) logits
        # target: (N,) class indices
        log_softmax = input.log_softmax(dim=1)
        nll_loss = -log_softmax.gather(1, target.unsqueeze(1))
        return self._reduce(nll_loss.squeeze(1))
```

**Binary Cross Entropy Loss**
```python
class BCELoss(Loss):
    """Binary Cross Entropy loss."""
    
    def forward(self, input, target):
        # input: probabilities [0, 1]
        # target: binary labels {0, 1}
        eps = 1e-8  # For numerical stability
        loss = -(target * input.clamp(eps, 1-eps).log() + 
                (1-target) * (1-input).clamp(eps, 1-eps).log())
        return self._reduce(loss)

class BCEWithLogitsLoss(Loss):
    """BCE loss with sigmoid for numerical stability."""
    
    def forward(self, input, target):
        # input: logits (any real number)
        # target: binary labels {0, 1}
        sigmoid_input = input.sigmoid()
        return BCELoss(self.reduction).forward(sigmoid_input, target)
```

### Required Helper Functions

**Softmax and Log-Softmax**
```python
class Softmax(Function):
    @staticmethod
    def forward(ctx, input, dim=-1):
        # Subtract max for numerical stability
        max_vals = input.max(dim=dim, keepdim=True)
        shifted = input - max_vals
        exp_vals = shifted.exp()
        sum_exp = exp_vals.sum(dim=dim, keepdim=True)
        result = exp_vals / sum_exp
        ctx.save_for_backward(result)
        ctx.dim = dim
        return result
    
    @staticmethod
    def backward(ctx, grad_output):
        softmax_output, = ctx.saved_tensors
        # Gradient computation for softmax
        return grad_softmax(softmax_output, grad_output, ctx.dim)
```

## 🧪 Testing Requirements

- [ ] Test all loss functions with different input shapes
- [ ] Test gradient computation with numerical differentiation
- [ ] Test reduction modes ('mean', 'sum', 'none')
- [ ] Test numerical stability with extreme values
- [ ] Test with different batch sizes
- [ ] Test edge cases (empty tensors, single samples)
- [ ] Integration tests with neural network training
- [ ] Performance benchmarks against PyTorch
- [ ] Test class weights and ignore_index functionality

## 📁 Files to Create/Modify

- `grad/nn/loss.py` - Loss function implementations
- `grad/nn/__init__.py` - Add loss function imports
- `grad/autograd/ops.py` - Add softmax, log_softmax operations
- `tests/test_loss_functions.py` - Comprehensive loss function tests
- `tests/test_loss_gradients.py` - Gradient verification tests
- `examples/loss_usage.py` - Usage examples

## 🔗 Dependencies

- Requires: Tensor operations (#06) - exp, log, sigmoid
- Requires: Enhanced sum/mean (#07) - for reduction operations  
- Requires: Neural network layers (#08) - Module base class
- Requires: Autograd system (#01, #03) - gradient computation
- Blocks: Complete training pipeline (#11)
- Blocks: Optimizer implementations (#10)

### Basic Usage
```python
from grad.nn import MSELoss, CrossEntropyLoss, BCELoss

# Regression with MSE
criterion = MSELoss()
predictions = Tensor([[1.0, 2.0], [3.0, 4.0]], requires_grad=True)
targets = Tensor([[1.5, 2.5], [2.5, 3.5]])
loss = criterion(predictions, targets)
loss.backward()

# Classification with Cross Entropy
criterion = CrossEntropyLoss()
logits = Tensor([[2.0, 1.0, 0.1], [0.5, 2.5, 1.0]], requires_grad=True)
targets = Tensor([0, 1])  # Class indices
loss = criterion(logits, targets)

# Binary classification
criterion = BCEWithLogitsLoss()
logits = Tensor([[0.5], [-1.2], [2.1]], requires_grad=True)
targets = Tensor([[1.0], [0.0], [1.0]])
loss = criterion(logits, targets)
```

### Training Integration
```python
# Neural network training example
model = Sequential(Linear(10, 5), ReLU(), Linear(5, 3))
criterion = CrossEntropyLoss()
optimizer = SGD(model.parameters(), lr=0.01)

for epoch in range(100):
    # Forward pass
    output = model(x_train)
    loss = criterion(output, y_train)
    
    # Backward pass
    model.zero_grad()
    loss.backward()
    optimizer.step()
    
    if epoch % 10 == 0:
        print(f'Epoch {epoch}, Loss: {loss.item():.4f}')
```

### Custom Loss Function
```python
class HuberLoss(Loss):
    """Huber loss for robust regression."""
    
    def __init__(self, delta=1.0, reduction='mean'):
        super().__init__(reduction)
        self.delta = delta
    
    def forward(self, input, target):
        diff = (input - target).abs()
        quadratic = diff.clamp(max=self.delta)
        linear = diff - quadratic
        loss = 0.5 * quadratic**2 + self.delta * linear
        return self._reduce(loss)
```

## 🚨 Edge Cases to Handle

- Numerical stability in log operations (add epsilon)
- Overflow in exponential functions (subtract max in softmax)
- Division by zero in probability computations
- Invalid target indices in CrossEntropyLoss
- NaN and infinity propagation
- Empty batch handling
- Very large or small logit values
- Class imbalance with weighted losses

## 🔧 Performance Considerations

- Numerically stable implementations (log-sum-exp trick)
- Memory-efficient softmax computation
- Avoid unnecessary intermediate tensors
- Efficient reduction operations
- Consider future GPU implementations
- Vectorized operations where possible

## 🏷️ Labels

- `enhancement`
- `loss-functions`
- `neural-networks`
- `high-priority`
- `training`

## 📈 Priority

**High** - Essential for neural network training and optimization

## 🔗 Related Issues

- Depends on: Tensor operations (#06)
- Depends on: Enhanced sum operation (#07)
- Depends on: Neural network layers (#08)
- Depends on: Autograd system (#01, #03)
- Blocks: Optimizers (#10)
- Blocks: Complete training pipeline (#11)
